File: C#/01KnapsackDP.cs
using System;
using System.Diagnostics;

class Program
{
    static void Main()
    {
        Console.Write("Enter number of items: ");
        int n = int.Parse(Console.ReadLine());

        int[] weight = new int[n];
        int[] value = new int[n];

        Console.WriteLine("Enter items as: <weight> <value>");
        for (int i = 0; i < n; i++)
        {
            string[] parts = Console.ReadLine().Split();
            weight[i] = int.Parse(parts[0]);
            value[i] = int.Parse(parts[1]);
        }

        Console.Write("Enter knapsack capacity: ");
        int capacity = int.Parse(Console.ReadLine());

        int[,] dp = new int[n + 1, capacity + 1];

        Stopwatch sw = Stopwatch.StartNew();

        // 0/1 Knapsack DP
        for (int i = 1; i <= n; i++)
        {
            for (int w = 0; w <= capacity; w++)
            {
                if (weight[i - 1] <= w)
                {
                    int take = value[i - 1] + dp[i - 1, w - weight[i - 1]];
                    int skip = dp[i - 1, w];
                    dp[i, w] = Math.Max(take, skip);
                }
                else
                {
                    dp[i, w] = dp[i - 1, w];
                }
            }
        }

        sw.Stop();

        Console.WriteLine("\nMaximum value achievable: " + dp[n, capacity]);
        Console.WriteLine("Execution time: " + sw.Elapsed.TotalSeconds + " seconds");

        // Optional: reconstruct selected items
        Console.WriteLine("\nSelected items (0-based index):");
        int remainingCapacity = capacity;
        for (int i = n; i > 0; i--)
        {
            if (dp[i, remainingCapacity] != dp[i - 1, remainingCapacity])
            {
                Console.WriteLine(
                    "Item " + (i - 1) +
                    " (weight=" + weight[i - 1] +
                    ", value=" + value[i - 1] + ")"
                );
                remainingCapacity -= weight[i - 1];
            }
        }
    }
}

File: Rust/01KnapsackDP.rs
use std::io::{self, Read};
use std::time::Instant;

fn knapsack_01(weights: &Vec<usize>, values: &Vec<i64>, capacity: usize) -> i64 {
    let mut dp = vec![0_i64; capacity + 1];

    for i in 0..weights.len() {
        let w = weights[i];
        let v = values[i];
        // Traverse backwards to ensure 0/1 behavior
        for c in (w..=capacity).rev() {
            let candidate = dp[c - w] + v;
            if candidate > dp[c] {
                dp[c] = candidate;
            }
        }
    }
    dp[capacity]
}

fn main() {
    // Read all input
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut it = input.split_whitespace();

    println!("Enter number of items:");
    let n: usize = match it.next() {
        Some(v) => v.parse().unwrap(),
        None => return,
    };

    let mut weights = Vec::with_capacity(n);
    let mut values = Vec::with_capacity(n);

    println!("Enter items as: <weight> <value>:");
    for _ in 0..n {
        let w: usize = it.next().unwrap().parse().unwrap();
        let v: i64 = it.next().unwrap().parse().unwrap();
        weights.push(w);
        values.push(v);
    }

    println!("Enter knapsack capacity:");
    let capacity: usize = it.next().unwrap().parse().unwrap();

    let start = Instant::now();
    let max_value = knapsack_01(&weights, &values, capacity);
    let duration = start.elapsed();

    println!("\nMaximum value achievable: {}", max_value);
    println!(
        "Execution time: {:.9} seconds",
        duration.as_secs_f64()
    );
}
